internal/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -0,0 +1,390 @@
+	}
+	defer conn.Release()
+
+	tag, err := conn.Exec(context.Background(), sqlCancelJob, id)
+	if err != nil {
+		return fmt.Errorf("error canceling job %s: %v", id, err)
+	}
+
+	if tag.RowsAffected() != 1 {
+		return jobqueue.ErrNotRunning
+	}
+
+	return nil
+}
+
+func (q *dbJobQueue) JobStatus(id uuid.UUID) (result json.RawMessage, queued, started, finished time.Time, canceled bool, deps []uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	var sp, fp *time.Time
+	var rp pgtype.JSON
+	err = conn.QueryRow(context.Background(), sqlQueryJobStatus, id).Scan(&rp, &queued, &sp, &fp, &canceled)
+	if err != nil {
+		return
+	}
+	if sp != nil {
+		started = *sp
+	}
+	if fp != nil {
+		finished = *fp
+	}
+	if rp.Status != pgtype.Null {
+		result = rp.Bytes
+	}
+
+	deps, err = q.jobDependencies(context.Background(), conn, id)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Job returns all the parameters that define a job (everything provided during Enqueue).
+func (q *dbJobQueue) Job(id uuid.UUID) (jobType string, args json.RawMessage, dependencies []uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	err = conn.QueryRow(context.Background(), sqlQueryJob, id).Scan(&jobType, &args, nil, nil)
+	if err == pgx.ErrNoRows {
+		err = jobqueue.ErrNotExist
+		return
+	} else if err != nil {
+		return
+	}
+
+	dependencies, err = q.jobDependencies(context.Background(), conn, id)
+	return
+}
+
+// Find job by token, this will return an error if the job hasn't been dequeued
+func (q *dbJobQueue) IdFromToken(token uuid.UUID) (id uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error establishing connection: %v", err)
+	}
+	defer conn.Release()
+
+	err = conn.QueryRow(context.Background(), sqlQueryId, token).Scan(&id)
+	if err == pgx.ErrNoRows {
+		return uuid.Nil, jobqueue.ErrNotExist
+	} else if err != nil {
+		return uuid.Nil, fmt.Errorf("Error retrieving id: %v", err)
+	}
+
+	return
+}
+
+// Get a list of tokens which haven't been updated in the specified time frame
+func (q *dbJobQueue) Heartbeats(olderThan time.Duration) (tokens []uuid.UUID) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	rows, err := conn.Query(context.Background(), sqlQueryHeartbeats, olderThan.String())
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var t uuid.UUID
+		_ = rows.Scan(&t) // TODO check err
+		tokens = append(tokens, t)
+	}
+
+	return
+}
+
+// Reset the last heartbeat time to time.Now()
+func (q *dbJobQueue) RefreshHeartbeat(token uuid.UUID) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return
+	}
+	defer conn.Release()
+
+	_, _ = conn.Exec(context.Background(), sqlRefreshHeartbeat, token)
```

Comment:
Missing error check

Reply:
Hm, do you think we should change the signature of these heartbeat functions to relay errors to the worker server so the server can decide what to do with them?

Reply:
It would probably just log them because there's nothing else to do, right? Maybe we should just log the errors here and expand the signature only if need to.

Reply:
agreed ya